cpp/src/arrow/compute/exec/filter_node.cc

```diff
@@ -113,7 +131,22 @@ class FilterNode : public ExecNode {
 
   void InputFinished(ExecNode* input, int total_batches) override {
     DCHECK_EQ(input, inputs_[0]);
+    auto executor = plan()->exec_context()->executor();
+
+    for (auto&& task : tasks_) {
```

Comment:
InputFinished (despite its name) can be called before InputReceived (it's less of an "input is done" and more of "this is how many inputs to expect"). So this won't necessarily work; you'll want something like the AtomicCounter pattern used in other exec nodes.

Reply:
Thanks  @lidavidm  I am using AtomicCounters and there is not more race condition issues